src/goto-instrument/synthesizer/expr_enumerator.cpp

```diff
@@ -0,0 +1,375 @@
+
+  std::list<partitiont> result;
+  while(v != end)
+  {
+    // Construct the parititon for current bit vector and add it to `result`
+    result.emplace_back(from_bits_to_partition(v, n));
+
+    // https://graphics.stanford.edu/~seander/bithacks.html#NextBitPermutation
+    // Compute the lexicographically next bit permutation.
+    std::size_t t = (v | (v - 1)) + 1;
+    v = t | ((((t & -t) / (v & -v)) >> 1) - 1);
+  }
+  result.emplace_back(from_bits_to_partition(v, n));
+
+  return result;
+}
+
+bool binary_functional_enumeratort::is_commutative(const irep_idt &op) const
+{
+  return op_id == ID_equal || op_id == ID_plus || op_id == ID_notequal ||
+         op_id == ID_or || op_id == ID_and || op_id == ID_xor ||
+         op_id == ID_bitand || op_id == ID_bitor || op_id == ID_bitxor ||
+         op_id == ID_mult;
+}
+
+bool binary_functional_enumeratort::is_equivalence_class_representation(
+  const expr_listt &exprs) const
+{
+  // When the two sub-enumerators are exchangeable---they enumerate the same
+  // set of expressions---, and the operator is commutative, `exprs` is a
+  // representation if its sub-expressions are sorted.
+  if(
+    is_exchangeable && is_commutative(op_id) &&
+    from_expr(exprs.front()) > from_expr(exprs.back()))
+    return false;
+
+  /// We are not sure if `exprs` is represented by some other tuple.
+  return true;
+}
+
+exprt binary_functional_enumeratort::instantiate(const expr_listt &exprs) const
+{
+  INVARIANT(
+    exprs.size() == 2,
+    "number of arugment should be 2: " + integer2string(exprs.size()));
+  if(op_id == ID_equal)
+    return equal_exprt(exprs.front(), exprs.back());
+  if(op_id == ID_le)
+    return less_than_or_equal_exprt(exprs.front(), exprs.back());
+  if(op_id == ID_lt)
+    return less_than_exprt(exprs.front(), exprs.back());
+  if(op_id == ID_gt)
+    return greater_than_exprt(exprs.front(), exprs.back());
+  if(op_id == ID_ge)
+    return greater_than_or_equal_exprt(exprs.front(), exprs.back());
+  if(op_id == ID_and)
+    return and_exprt(exprs.front(), exprs.back());
+  if(op_id == ID_or)
+    return or_exprt(exprs.front(), exprs.back());
+  if(op_id == ID_plus)
+    return plus_exprt(exprs.front(), exprs.back());
+  if(op_id == ID_minus)
+    return minus_exprt(exprs.front(), exprs.back());
+  if(op_id == ID_notequal)
+    return notequal_exprt(exprs.front(), exprs.back());
+  return binary_exprt(exprs.front(), op_id, exprs.back());
+}
+
+expr_sett alternatives_enumeratort::enumerate(const std::size_t size) const
+{
+  expr_sett result;
+  for(const auto &enumerator : sub_enumerators)
+  {
+    for(const auto &e : enumerator->enumerate(size))
+    {
+      result.insert(e);
+    }
+  }
+  return result;
+}
+
+expr_sett
+recursive_enumerator_placeholdert::enumerate(const std::size_t size) const
+{
+  INVARIANT(factory.productions_map.count(identifier), "No nonterminal found.");
+  alternatives_enumeratort actual_enumerator(
+    factory.productions_map.at(identifier), ns);
```

Comment:
This entails two lookups of the same element. Perhaps use `.find` and then and `INVARIANT` over the resulting iterator?